### PR TITLE
For discussion: master election using etcd

### PR DESCRIPTION
**Pull request by daviddrysdale**
Best reviewed commit-by-commit.

Some significant project-setup changes included: vendors and submodules and etcd, oh my.

### CODE REVIEW

File: util/tracker.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	mt.mu.RLock()
+	defer mt.mu.RUnlock()
+	return mt.masterCount
+}
+
+// Held returns a (sorted) list of the IDs for which we are currently master.
+func (mt *MasterTracker) Held() []int64 {
+	mt.mu.RLock()
+	defer mt.mu.RUnlock()
+	ids := make([]int64, 0, len(mt.masterFor))
```
Comment (AlCutter) [inline util/tracker.go]: `len(mt.masterFor)` => `mt.masterCount`?

File: util/election.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
```
Comment (AlCutter) [inline util/election.go]: `import "context"`

File: util/election.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"context"
+)
+
+// MasterElection provides operations for determining if a local instance is the current
+// master for a particular election.
+type MasterElection interface {
+	// Start kicks off the process of mastership election.
+	Start(ctx context.Context) error
```
Comment (AlCutter) [inline util/election.go]: I think you can drop the parameter names here

File: util/election.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	// WaitForMastership blocks until the current instance is master for this election.
+	WaitForMastership(ctx context.Context) error
+	// IsMaster returns whether the current instance is master.
+	IsMaster(ctx context.Context) (bool, error)
+	// Resign releases mastership, and re-joins the election.
+	Resign(ctx context.Context) error
```
Comment (AlCutter) [inline util/election.go]: Maybe name it `ResignAndRejoin`?

File: util/election.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// NoopElection is a stub implementation that tells every instance that it is master.
+type NoopElection struct {
+	treeID     int64
+	instanceID string
+}
+
+// Start kicks off the process of mastership election.
+func (emm *NoopElection) Start(ctx context.Context) error {
```
Comment (AlCutter) [inline util/election.go]: nit: emm?

File: util/etcd/election.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+// IsMaster returns whether the current instance is the master.
+func (eme *MasterElection) IsMaster(ctx context.Context) (bool, error) {
+	leader, err := eme.election.Leader(ctx)
+	if err != nil {
+		return false, err
+	}
+	return leader == eme.instanceID, nil
+}
+
+// Resign relinquishes mastership.
```
Comment (AlCutter) [inline util/etcd/election.go]: The interface docs says this'll rejoin, but this comment is strangely silent on that bit :)

File: server/trillian_log_signer/main.go [hunk 7]
```diff
@@ -39,6 +42,9 @@ var (
 	numSeqFlag               = flag.Int("num_sequencers", 10, "Number of sequencer workers to run in parallel")
 	sequencerGuardWindowFlag = flag.Duration("sequencer_guard_window", 0, "If set, the time elapsed before submitted leaves are eligible for sequencing")
 	dumpMetricsInterval      = flag.Duration("dump_metrics_interval", 0, "If greater than 0, how often to dump metrics to the logs.")
+	forceMaster              = flag.Bool("force_master", false, "If true, assume master for all logs")
```
Comment (AlCutter) [inline server/trillian_log_signer/main.go]: woah, scary!

File: server/trillian_log_signer/main.go [hunk 8]
```diff
@@ -59,10 +65,24 @@ func main() {
 	}
 	defer db.Close()
 
+	ctx, cancel := context.WithCancel(context.Background())
+	go util.AwaitSignal(cancel)
+
+	hostname, _ := os.Hostname()
+	instanceID := fmt.Sprintf("%s.%d", hostname, os.Getpid())
```
Comment (AlCutter) [inline server/trillian_log_signer/main.go]: There's this btw: https://github.com/google/uuid

File: integration/ct_prep_test.sh [hunk 9]
```diff
@@ -22,10 +22,29 @@ go build ${GOFLAGS} ./examples/ct/ct_server/
...
+    ${ETCD_DIR}/etcd &
+    ETCD_PID=$!
+    ETCD_DB_DIR=default.etcd
+    set +e
+    waitForServerStartup ${ETCD_PORT}
+    set -e
+    SIGNER_ELECTION_OPTS="--etcd_servers=${ETCD_SERVER}"
+else
+    if  [[ ${LOG_SIGNER_COUNT} > 1 ]]; then
+        echo "*** Warning: running multiple signers with no etcd instance **"
```
Comment (AlCutter) [inline integration/ct_prep_test.sh]: Need an extra `*` on the end of the string, you anarchist!

File: server/log_operation_manager.go [hunk 10]
```diff
@@ -54,27 +62,142 @@ type LogOperationInfo struct {
...
+		if err := er.election.WaitForMastership(ctx); err != nil {
+			glog.Errorf("%d: er.election.WaitForMastership() failed: %v", er.logID, err)
+			return
+		}
+		glog.V(1).Infof("%d: Now, I am the master", er.logID)
+		er.tracker.Set(er.logID, true)
+		masterSince := time.Now()
+
+		// While-master loop
+		checkInterval := er.info.MasterCheckInterval
```
Comment (AlCutter) [inline server/log_operation_manager.go]: you could do this clamp-upwards checks up top next to the pre-election one (and possibly move them out into a fixMyInputs type func which gets called at the top of this func

File: server/log_operation_manager.go [hunk 11]
```diff
@@ -54,27 +62,142 @@ type LogOperationInfo struct {
...
+				glog.Errorf("%d: failed to resign mastership", er.logID)
+			}
+		}
+	}
+}
+
+func (er *electionRunner) shouldResign(masterSince time.Time) bool {
+	now := er.info.TimeSource.Now()
+	duration := now.Sub(masterSince)
+	holdFor := er.info.MasterHoldInterval
```
Comment (AlCutter) [inline server/log_operation_manager.go]: and this one

File: server/log_operation_manager.go [hunk 12]
```diff
@@ -92,11 +215,60 @@ func (l LogOperationManager) getLogIDs(ctx context.Context) ([]int64, error) {
...
+	allIDs, err := l.getLogIDs(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve full list of log IDs: %v", err)
+	}
+	logIDs, err := l.masterFor(ctx, allIDs)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to determine log IDs we're master for: %v", err)
 	}
+	l.heldMutex.Lock()
```
Comment (AlCutter) [inline server/log_operation_manager.go]: could move this block out into an `updateMasterIDs(logIDs)` (or something func), that'd trim this func size down and let you do the standard immediate `defer` thing to release the mutex lock.

File: util/timesource_test.go [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+	"testing"
+	"time"
+)
+
+func TestFakeTimeSource(t *testing.T) {
+	date1 := time.Date(1970, 9, 19, 12, 00, 00, 00, time.UTC)
+	date2 := time.Date(2007, 7, 7, 11, 35, 00, 00, time.UTC)
+	fake := NewFakeTimeSource(date1)
+
+	var ts TimeSource
```
Comment (daviddrysdale) [inline util/timesource_test.go]: Done.

File: util/tracker.go [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+	return &MasterTracker{masterFor: mf}
+}
+
+// Set changes the tracked mastership status for the given id.  This method should
+// be called exactly once for each state transition.
+func (mt *MasterTracker) Set(id int64, val bool) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	existing, ok := mt.masterFor[id]
+	if ok && val != !existing {
```
Comment (daviddrysdale) [inline util/tracker.go]: Done.

File: util/tracker.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+		glog.Errorf("toggle masterFor[%d] from %v to %v!", id, existing, val)
```
Comment (daviddrysdale) [inline util/tracker.go]: I think returning an error would make life much harder for the caller -- it's not really going to be able to do anything other than log it.  I have downgraded to a `Warningf` -- the function can still do what it's asked in this case, it's just a breach of an expected precondition.

File: util/tracker.go (hunk 1 above)
Comment (daviddrysdale) [inline util/tracker.go]: Done.

File: util/election.go (hunk 2 above)
Comment (daviddrysdale) [inline util/election.go]: Done.

File: util/election.go (hunk 3 above)
Comment (daviddrysdale) [inline util/election.go]: Done.

File: util/election.go (hunk 4 above)
Comment (daviddrysdale) [inline util/election.go]: Went for `ResignAndRestart` to echo `Start`

File: util/election.go (hunk 5 above)
Comment (daviddrysdale) [inline util/election.go]: Done.

File: util/etcd/election.go (hunk 6 above)
Comment (daviddrysdale) [inline util/etcd/election.go]: Done.

File: server/log_operation_manager_test.go [hunk 16]
```diff
@@ -129,7 +150,14 @@ func TestLogOperationManagerPassesIDs(t *testing.T) {
 	mockLogOp.EXPECT().ExecutePass(gomock.Any(), logID1, infoMatcher)
 	mockLogOp.EXPECT().ExecutePass(gomock.Any(), logID2, infoMatcher)
 
-	lom := NewLogOperationManager(registry, 50, 1, time.Second, fakeTimeSource, mockLogOp)
+	info := LogOperationInfo{
```
Comment (daviddrysdale) [inline server/log_operation_manager_test.go]: Done.

File: server/trillian_log_signer/main.go (hunk 7 above)
Comment (daviddrysdale) [inline server/trillian_log_signer/main.go]: Yeah, but there could well be situations where it's very helpful to be able to spin up a signer that doesn't depend on etcd to get on with the job...

File: server/trillian_log_signer/main.go (hunk 8 above)
Comment (daviddrysdale) [inline server/trillian_log_signer/main.go]: Maybe later: "This package is currently in development and the API may not be stable."

File: integration/ct_prep_test.sh (hunk 9 above)
Comment (daviddrysdale) [inline integration/ct_prep_test.sh]: Done.

File: server/log_operation_manager.go (hunk 12 above)
Comment (daviddrysdale) [inline server/log_operation_manager.go]: Done.

File: server/log_operation_manager.go (hunk 10 above)
Comment (daviddrysdale) [inline server/log_operation_manager.go]: Done.

File: server/log_operation_manager.go (hunk 11 above)
Comment (daviddrysdale) [inline server/log_operation_manager.go]: Done.

**Comment by codingllama** [comment]
Could we have the vendoring done in a separate PR? Also, in regards to it, have you considered using `git subtree` instead of `git submodule`?

https://www.atlassian.com/blog/git/alternatives-to-git-submodule-git-subtree

**Comment by daviddrysdale** [comment]
I can't put the vendoring into a separate PR without having broken CI, so I'd prefer to leave it where it is.  (BTW, I'm expecting to my normal thing of rebase-and-merge so the Git history will be clear in future.)

And yes, I plan to shift to subtree rather than submodule.

**Comment by daviddrysdale** [comment]
Shifted to use subtrees, rebased and folded in code review markups.  One extra commit, updating README to describe subtree use.

File: README.md [hunk 17]
```diff
@@ -158,6 +159,35 @@ and run the following:
...
+upstream repository does not guarantee back-compatibility between the tip
+`master` branch and the current stable release).  These external codebases are
+included as Git
+[subtrees](https://github.com/git/git/blob/master/contrib/subtree/git-subtree.txt).
+
+To update the code in one of these subtrees, perform steps like:
+
+```bash
+# Add master repo for upstream code as a Git remote.
+git remote add vendor-xyzzy https://github.com/orgname/xyzzy
```
Comment (AlCutter) [inline README.md]: [Rush](https://www.youtube.com/watch?v=_mR0tzt9XaM)

**Pull request merged**